nifi-nar-bundles/nifi-kite-bundle/nifi-kite-processors/src/main/java/org/apache/nifi/processors/kite/InferAvroSchema.java

```diff
@@ -0,0 +1,454 @@
+        return properties;
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        final FlowFile original = session.get();
+        if (original == null) {
+            return;
+        }
+
+        try {
+
+            final AtomicReference<String> avroSchema = new AtomicReference<>();
+            switch (context.getProperty(INPUT_CONTENT_TYPE).getValue()) {
+                case USE_MIME_TYPE:
+                    avroSchema.set(inferAvroSchemaFromMimeType(original, context, session).get());
+                    break;
+                case JSON_CONTENT:
+                    avroSchema.set(inferAvroSchemaFromJSON(original, context, session).get());
+                    break;
+                case CSV_CONTENT:
+                    avroSchema.set(inferAvroSchemaFromCSV(original, context, session).get());
+                    break;
+                default:
+                    //Shouldn't be possible but just in case
+                    session.transfer(original, REL_UNSUPPORTED_CONTENT);
+                    break;
+            }
+
+
+            if (StringUtils.isNotEmpty(avroSchema.get())) {
+
+                String destination = context.getProperty(SCHEMA_DESTINATION).getValue();
+                FlowFile avroSchemaFF = null;
+
+                switch (destination) {
+                    case DESTINATION_ATTRIBUTE:
+                        avroSchemaFF = session.putAttribute(session.clone(original), AVRO_SCHEMA_ATTRIBUTE_NAME, avroSchema.get());
+                        //Leaves the original CoreAttributes.MIME_TYPE in place.
+                        break;
+                    case DESTINATION_CONTENT:
+                        avroSchemaFF = session.write(session.create(), new OutputStreamCallback() {
+                            @Override
+                            public void process(OutputStream out) throws IOException {
+                                out.write(avroSchema.get().getBytes());
+                            }
+                        });
+                        avroSchemaFF = session.putAttribute(avroSchemaFF, CoreAttributes.MIME_TYPE.key(), AVRO_MIME_TYPE);
+                        break;
+                    default:
+                        break;
+                }
+
+                //Transfer the sessions.
+                avroSchemaFF = session.putAttribute(avroSchemaFF, CoreAttributes.FILENAME.key(), (original.getAttribute(CoreAttributes.FILENAME.key()) + AVRO_FILE_EXTENSION));
+                session.transfer(avroSchemaFF, REL_SUCCESS);
+            } else {
+                //If the avroSchema is null then the content type is unknown and therefore unsupported
+                session.transfer(session.clone(original), REL_UNSUPPORTED_CONTENT);
```

Comment:
If transferring to "unsupported content" I would recommend transferring the original FlowFile to 'unsupported content' and transferring nothing to 'original' - this is the more typical usage pattern.